### PR TITLE
Building a Simple Nest.js Microservice Application using Redis as the Transporter

### LINKED ISSUE #2680: Building a Nest.js microservice application using Redis as the transporter
We ask that you please be patient as our team works through approving and publishing all articles/tutorials in a timely manner. Allow 1-3 days for a topic to be reviewed and/or approved - allow 3-7 days for an articles to be reviewed and/or published.

Be sure to visit our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for tools, resources, and example articles to go over.

**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**NOTE: We tend to stray away or tend not to publish reviews/comparisons of commercial product offerings.**

*We're looking for the first 2-3 paragraphs of the article that appropriately summarize what your article will be about.*

### Proposed title of the article
Your title should be descriptive of the article/tutorial. Be Specific. *Use keyword research to gain your article a higher ranking.*

### Introduction paragraph (2-3 paragraphs):

Nest.js is a JavaScript framework for building efficient and scalable hybrid static and server-side applications. Nest.js is built on top of React to abstracts rendering complexity, giving you the flexibility to build scalable React applications. This hybrid enables your React apps to have additional functionality, such as server-side rendering. The Nest.js framework will allow you to build a react app but render the content in advance on the server-side. Traditional with React, you can only render on the client-side. Nest.js gives you the flexibility to choose which data fetching strategy fits your application as you grow and scale your application from a few pages to thousands. With server-side rendering, pages are rendered during the build-up phase. The content is served ahead of time to user or search engine bots. This way, they will only receive the fully rendered HTML. After receiving this initial page, client-side rendering takes over and works just like a traditional React app. This tutorial will teach you how to create a Nest.js Microservice application with Redis.

##### An Overview of a microservice

Microservice is an architectural design that allows you to create small modules/components of an application that can scale on demand. Each component is packaged independently and can communicate with each other via APIs. Nest.js comes with built-in support for microservice architecture and multiple transport layers. These transport layers are in charge of transferring messages between microservice instances. We can use Redis as a transport layer protocol for end-to-end connectivity to allow flow control and segmentation between different microservices. As a result, every microservice can connect to different channels, and every channel can have various microservices connected to it.

### Key takeaways: 

What is Nest.js
An Overview of a microservice
Setting Redis and creating client and server apps
Setting Nest.js microservice in both apps
Sending a message from the client to the server using Redis as the transporter

### References: 
Please list links to any published content/research that you intend to use to support/guide this article.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by kimkimani**
### Article checklist

**ATTENTION: In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.**
      
      
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
Fixes #2680

### CODE REVIEW

File: content/articles/building-a-simple-nestjs-microservice-application-using-redis-as-the-transporter/index.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+### An Overview of a microservice
+
+Microservice is an architectural design that allows you to create small modules/components of an application that can scale on demand. Each component is packaged independently and can communicate with each other via APIs. Nest.js comes with built-in support for microservice architecture and multiple transport layers. These transport layers are in charge of transferring messages between microservice instances. We can use Redis as a transport layer protocol for end-to-end connectivity to allow flow control and segmentation between different microservices. As a result, every microservice can connect to different channels, and every channel can have various microservices connected to it. Let’s now build a Nest.js microservice application using Redis as the transporter.
+
+### Prerequisites
+
+To follow along with this guide, the following prerequisites will be essential;
+
+- We are going to buddle Nest.js CLI using a Node.js package. Ensure you have Node.js installed so that you can have NPM on your computer. You can download the Node.js installer here.
```
Comment (zolomohan) [inline content/articles/building-a-simple-nestjs-microservice-application-using-redis-as-the-transporter/index.md:15]: Please add the link to Node.js Download Page.

File: content/articles/building-a-simple-nestjs-microservice-application-using-redis-as-the-transporter/index.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+- Nest.js is bunddled and backed by Typescript. So a basic knowledge of Typescript is requried.
+
+- To implement a microservice architecture, we are going to use Redis as the transporter. Make sure Redis is installed on your computer.
```
Comment (zolomohan) [inline content/articles/building-a-simple-nestjs-microservice-application-using-redis-as-the-transporter/index.md:19]: Please add the link to Redis download page.

**Review by zolomohan — CHANGES_REQUESTED** [review]

File: content/articles/building-a-simple-nestjs-microservice-application-using-redis-as-the-transporter/index.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+```bash
+npm start
+```
+
+on one terminal, open another, and run;
+
+```bash
+npm start new-app
+```
```
Comment (zolomohan) [inline content/articles/building-a-simple-nestjs-microservice-application-using-redis-as-the-transporter/index.md:126]: Please instruct the reader to start the redis-server as well.

**Review by zolomohan — CHANGES_REQUESTED** [review]

File: content/articles/building-a-simple-nestjs-microservice-application-using-redis-as-the-transporter/index.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+```bash
+nest new nest-ms
+```
+
+We need to turn our newly created project into a monorepo. A monorepo refers to a single repository hosting different microservices. To generate the monorepo in the app, we run the following command.
+
+```bash
+nest generate app new_app
+```
```
Comment (zolomohan) [inline content/articles/building-a-simple-nestjs-microservice-application-using-redis-as-the-transporter/index.md:41]: It's a bit unclear if I should run this command inside the nest-ms folder or outside. Please make it clear here.

**Review by zolomohan — CHANGES_REQUESTED** [review]

**Comment by zolomohan** [comment]
@kimkimani any updates?

**Comment by zolomohan** [comment]
@kimkimani?

**Comment by kimkimani** [comment]
> @kimkimani any updates?

Check the updates @zolomohan

**Comment by hectorkambow** [comment]
plagiarism check done

**Comment by WanjaMIKE** [comment]
Will be conducting a final review.

**Comment by WanjaMIKE** [comment]
Changing the branch name from `building-a-simple-nestjs-microservice-application-using-redis-as-the-transporter` to `building-nest-js-application-using-redis` for better SEO.

**Comment by hectorkambow** [comment]
Great work gents @kimkimani 

Very easy to follow along 👍 

cc @zolomohan & @WanjaMIKE 
 🚀

**Pull request merged**